circuits/cpp/barretenberg/cpp/src/barretenberg/stdlib/recursion/honk/verifier/ultra_recursive_verifier.cpp

```diff
@@ -0,0 +1,206 @@
+#include "barretenberg/stdlib/recursion/honk/verifier/ultra_recursive_verifier.hpp"
+#include "barretenberg/honk/flavor/standard.hpp"
+#include "barretenberg/honk/transcript/transcript.hpp"
+#include "barretenberg/honk/utils/grand_product_delta.hpp"
+#include "barretenberg/honk/utils/power_polynomial.hpp"
+#include "barretenberg/numeric/bitop/get_msb.hpp"
+
+namespace proof_system::plonk::stdlib::recursion::honk {
+template <typename Flavor>
+UltraRecursiveVerifier_<Flavor>::UltraRecursiveVerifier_(Builder* builder,
+                                                         std::shared_ptr<typename Flavor::VerificationKey> verifier_key)
+    : key(verifier_key)
+    , builder(builder)
+{}
+
+template <typename Flavor>
+UltraRecursiveVerifier_<Flavor>::UltraRecursiveVerifier_(UltraRecursiveVerifier_&& other) noexcept
+    : key(std::move(other.key))
+    , pcs_verification_key(std::move(other.pcs_verification_key))
+{}
+
+template <typename Flavor>
+UltraRecursiveVerifier_<Flavor>& UltraRecursiveVerifier_<Flavor>::operator=(UltraRecursiveVerifier_&& other) noexcept
+{
+    key = other.key;
+    pcs_verification_key = (std::move(other.pcs_verification_key));
+    commitments.clear();
+    pcs_fr_elements.clear();
+    return *this;
+}
+
+/**
+ * @brief This function verifies an Ultra Honk proof for given program settings.
+ *
+ */
+template <typename Flavor> std::array<typename Flavor::GroupElement, 2> UltraRecursiveVerifier_<Flavor>::verify_proof(const plonk::proof& proof)
+{
+    using FF = typename Flavor::FF;
+    using GroupElement = typename Flavor::GroupElement;
+    using Commitment = typename Flavor::Commitment;
+    using Sumcheck = ::proof_system::honk::sumcheck::SumcheckVerifier<Flavor>;
+    using Curve = typename Flavor::Curve;
+    using Gemini = ::proof_system::honk::pcs::gemini::GeminiVerifier_<Curve>;
+    using Shplonk = ::proof_system::honk::pcs::shplonk::ShplonkVerifier_<Curve>;
+    using PCS = typename Flavor::PCS; // note: This can only be KZG
+    using VerifierCommitments = typename Flavor::VerifierCommitments;
+    using CommitmentLabels = typename Flavor::CommitmentLabels;
+    using RelationParams = ::proof_system::honk::sumcheck::RelationParameters<FF>;
+
+    RelationParams relation_parameters;
+
+    size_t prev_num_gates = builder->get_num_gates();
+
+    transcript = Transcript<Builder>{ builder, proof.proof_data };
+    
+    auto commitments = VerifierCommitments(key);
+    auto commitment_labels = CommitmentLabels();
+
+    const auto circuit_size = transcript.template receive_from_prover<uint32_t>("circuit_size");
+    const auto public_input_size = transcript.template receive_from_prover<uint32_t>("public_input_size");
+    const auto pub_inputs_offset = transcript.template receive_from_prover<uint32_t>("pub_inputs_offset");
+
+    // Extract native integer types for these basic quantities for use in subsequent operations
+    auto circuit_size_native = static_cast<size_t>(circuit_size.get_value());
```

Comment:
Went ahead and just ditched these in favor of using the static cast in place. Don't love that either but it might be more clear than defining these additional variables

Reply:
yeah it took me a bit to understand the point of `native` variables and I think the in-place cast (and maybe comments why they're there) might be better